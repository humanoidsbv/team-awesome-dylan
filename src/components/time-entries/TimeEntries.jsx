import React, { useState } from 'react';

import styles from './TimeEntries.module.css';
import TimeEntry from '../time-entry/TimeEntry';
import TimeEntryForm from '../time-entry-form/TimeEntryForm';
import TimeEntryHeading from '../time-entry-heading/TimeEntryHeading';
import timeEntriesMock from './timeEntries.json';

function TimeEntries() {
  const [timeEntries, setTimeEntries] = useState(timeEntriesMock);

  const createTimeEntry = newTimeEntry => {
    setTimeEntries([newTimeEntry, ...timeEntries]);
  };

  return (
    <div className={styles.timeEntriesContainer}>
      <TimeEntryForm createTimeEntry={createTimeEntry} />
      {timeEntries.map(
        ({ client, id, startTimestamp, endTimestamp }, index) => (
          <React.Fragment key={id}>
            {(index === 0 ||
              (index > 0 &&
                timeEntries[index - 1].startTimestamp.split(' ')[0] !==
                  startTimestamp.split(' ')[0])) && (
              // eslint-disable-next-line react/jsx-indent
              <TimeEntryHeading startDate={startTimestamp.slice(0, 11)} />
            )}
            <TimeEntry
              client={client}
              clientDuration={
                endTimestamp.slice(11, 13) - startTimestamp.slice(11, 13)
              }
              endTime={endTimestamp.slice(11, 16)}
              key={id}
              startTime={startTimestamp.slice(11, 16)}
            />
          </React.Fragment>
        )
      )}
    </div>
  );
}

export default TimeEntries;
