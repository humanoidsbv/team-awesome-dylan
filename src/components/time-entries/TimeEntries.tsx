import React, { useEffect, useState } from 'react';

import Button from '../../shared/components/button/Button';
import TimeEntry from '../time-entry/TimeEntry';
import TimeEntryForm from '../time-entry-form/TimeEntryForm';
import TimeEntryHeading from '../time-entry-heading/TimeEntryHeading';
import styles from './TimeEntries.module.css';
import { TimeEntriesProps, TimeEntryInterface } from '../../redux/time-entries/types';

function TimeEntries({
  clients,
  createTimeEntry,
  deleteTimeEntry,
  fetchClients,
  fetchTimeEntries,
  filterTimeEntriesByClient,
  timeEntries
}: TimeEntriesProps): React.ReactElement {
  const [isTimeEntryFormVisible, setTimeEntryFormVisibility] = useState(false);

  const toggleTimeEntryForm = (): void => {
    setTimeEntryFormVisibility(!isTimeEntryFormVisible);
  };

  useEffect(() => {
    fetchClients();
    fetchTimeEntries();
  }, []);

  const handleChange = (event): {} => filterTimeEntriesByClient(
    !event.target.value ? null : Number(event.target.value)
  );

  return (
    <div className={styles.timeEntriesContainer}>
      <Button disabled={isTimeEntryFormVisible} onClick={toggleTimeEntryForm}>
        + New Time Entry
      </Button>
      <TimeEntryForm
        clients={clients}
        createTimeEntry={(newTimeEntry): {} => createTimeEntry(newTimeEntry)}
        isTimeEntryFormVisible={isTimeEntryFormVisible}
        toggleTimeEntryForm={toggleTimeEntryForm}
      />
      <div className={styles.headerTimeEntries}>
        <span className={styles.headerTimeEntriesText}> Time Entries </span>
        <select
          className={styles.filterClientsSelector}
          id="filterSelect"
          onChange={handleChange}
        >
          <option value="">Filter by:</option>
          {clients.map(({ clientName, id }) => (
            <option key={id} value={id}>
              {clientName}
            </option>
          ))}
        </select>
      </div>
      {timeEntries.map(
        ({
          client, id, startTimestamp, stopTimestamp
        }: TimeEntryInterface, index) => {
          const startDate = new Date(startTimestamp).toDateString();
          const previousDate = index > 0
            ? new Date(timeEntries[index - 1].startTimestamp).toDateString()
            : '';
          return (
            <React.Fragment key={id}>
              {previousDate !== startDate && (
                // eslint-disable-next-line react/jsx-indent
                <TimeEntryHeading startTimestamp={startTimestamp} />
              )}
              <TimeEntry
                client={client}
                deleteTimeEntry={(timeEntryId): {} => deleteTimeEntry(timeEntryId)}
                id={id}
                startTimestamp={startTimestamp}
                stopTimestamp={stopTimestamp}
              />
            </React.Fragment>
          );
        }
      )}
    </div>
  );
}

export default TimeEntries;
