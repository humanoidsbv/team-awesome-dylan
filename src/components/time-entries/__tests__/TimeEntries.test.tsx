import * as React from 'react';
import { shallow } from 'enzyme';

import TimeEntries from '../TimeEntries';

it('matches snapshot', () => {
  const timeEntries = shallow(
    <TimeEntries
      clients ={[{
        branch:'test',
        clientName:'test',
        id: 123
      }]}
      timeEntries={[{
        id: 123,
        startTimestamp: 'bla',
        stopTimestamp: 'bla',
        client:{
          branch: '123',
          clientName: '123',
          id: 123,
          locality: '123',
          postalCode: '123',
          contactPerson: 'test',
          contactPhoneNumber: 'test'
        }
      }]}
      createTimeEntry={({}) => null}
      deleteTimeEntry={(id: number) => null}
      fetchTimeEntries={()=> null}
      fetchClients={()=> null}
      filterTimeEntriesByClient={(id: number) => null}
    />
  );
  expect(timeEntries).toMatchSnapshot();
});
