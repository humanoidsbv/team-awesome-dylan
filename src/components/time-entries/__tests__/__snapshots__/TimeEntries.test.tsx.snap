// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="timeEntriesContainer"
>
  <styled.button
    disabled={false}
    onClick={[Function]}
  >
    + New Time Entry
  </styled.button>
  <TimeEntryForm
    clients={
      Array [
        Object {
          "branch": "test",
          "clientName": "test",
          "id": 123,
        },
      ]
    }
    createTimeEntry={[Function]}
    isTimeEntryFormVisible={false}
    toggleTimeEntryForm={[Function]}
  />
  <div
    className="headerTimeEntries"
  >
    <span
      className="headerTimeEntriesText"
    >
       Time Entries 
    </span>
    <select
      className="filterClientsSelector"
      id="filterSelect"
      onChange={[Function]}
    >
      <option
        value=""
      >
        Filter by:
      </option>
      <option
        key="123"
        value={123}
      >
        test
      </option>
    </select>
  </div>
  <TimeEntryHeading
    startTimestamp="bla"
  />
  <TimeEntry
    client={
      Object {
        "branch": "123",
        "clientName": "123",
        "contactPerson": "test",
        "contactPhoneNumber": "test",
        "id": 123,
        "locality": "123",
        "postalCode": "123",
      }
    }
    deleteTimeEntry={[Function]}
    id={123}
    startTimestamp="bla"
    stopTimestamp="bla"
  />
</div>
`;
